ray-curator/ray_curator/stages/modules/filter.py

```diff
@@ -0,0 +1,500 @@
+
+        """
+
+        if self.processing_batch_size == 1:
+            return [self.process(tasks[0])]
+
+        if self.id_field is None:
+            msg = "id_field must be provided for batch_size > 1"
+            raise ValueError(msg)
+
+        id_field = self.id_field
+        id_to_batch = {}
+        dfs = []
+
+        for _, task in enumerate(tasks):
+            df = task.to_pandas().copy()
+            if id_field not in df.columns:
+                msg = f"Expected id_field '{id_field}' not found in batch {task.task_id}"
+                raise ValueError(msg)
+
+            for doc_id in df[id_field]:
+                id_to_batch[doc_id] = (task.task_id, task.dataset_name)
+
+            dfs.append(df)
+
+        # Combine into a single Pandas DataFrame
+        combined_df = pd.concat(dfs, ignore_index=True)
+
+        # Process the all DocumentBatch objects together as a single DocumentBatch
+        combined_batch = DocumentBatch(data=combined_df, task_id="batch_list", dataset_name="batch_list")
+        processed_batch = self.process(combined_batch)
+
+        if processed_batch is None:
+            return []
+
+        filtered_df = processed_batch.to_pandas()
+
+        # Group rows back into batches by id_field
+        grouped = {}
+        for _, row in filtered_df.iterrows():
+            doc_id = row[id_field]
+            if doc_id not in id_to_batch:
+                msg = f"Filtered document ID '{doc_id}' not found in original batches"
+                raise ValueError(msg)
+
+            task_id, dataset_name = id_to_batch[doc_id]
+            grouped.setdefault((task_id, dataset_name), []).append(row)
+
+        # Rebuild original DocumentBatch objects after filtering
+        result_batches = []
+        for (task_id, dataset_name), rows in grouped.items():
+            batch_df = pd.DataFrame(rows)
+            result_batches.append(
+                DocumentBatch(
+                    task_id=f"{task_id}_{self.name}", dataset_name=dataset_name, data=batch_df.reset_index(drop=True)
+                )
+            )
+
+        return result_batches
+
+    def process(self, batch: DocumentBatch) -> DocumentBatch | None:
+        """
+        Applies the filtering to a dataset
+
+        Args:
+            batch (DocumentBatch): The batch to apply the module to
+
+        Returns:
+            DocumentBatch: A batch with entries removed according to the filter
+
+        """
+        df = batch.to_pandas()
+        bool_mask = self.compute_filter_mask(df)
+        result_df = df[bool_mask]
+
+        if len(result_df) == 0:
+            print(f"All documents filtered out for batch {batch.task_id}")
+            return None
+
+        # Create output batch
+        return DocumentBatch(
+            task_id=f"{batch.task_id}_{self.name}",
+            dataset_name=batch.dataset_name,
+            data=result_df,
+        )
+
+
+@dataclass
+class ScoreFilter(ProcessingStage[DocumentBatch, DocumentBatch]):
```

Comment:
When do users use Score and Filter separately vs using ScoreFilter?

Reply:
Score is for if you do not want to filter but still want to use the `DocumentFilter` to add the column. Filter is for if the score column already exists and you want to filter by the column, using a filtering method from a `DocumentFilter`.